File: ChMS.Web/Controllers/ExpenseController.cs
using Microsoft.AspNetCore.Mvc;
using Chms.Application.Common.Interface;
using Chms.Domain.Entities;
using Microsoft.AspNetCore.Authorization;
using ChMS.Web.ViewModels;
using Expense = Chms.Domain.Entities.Expense;
using Chms.Domain.ViewModels.IncomeExpense;


// For more information on enabling Web API for empty projects, visit https://go.microsoft.com/fwlink/?LinkID=397860

namespace ChMS.Web.Controllers
{
    [Route("[controller]")]
    [ApiController]
    [Authorize]
    public class ExpenseController : ControllerBase
    {
        public readonly IExpenseService _service;
        public readonly IFileUploadService _uploadService;
        private readonly ICurrentUserService _currentUserService;
        public ExpenseController(IExpenseService service, IFileUploadService uploadService, ICurrentUserService currentUserService)
        {
            _service = service;
            _uploadService = uploadService;
            _currentUserService = currentUserService;
        }


        [HttpPost]
        public async Task<ActionResult<int>> Create([FromBody] AddExpense request)
        {
            try
            {
                if (request.Expense.Count == 0 || String.IsNullOrEmpty(request.Date.ToString()))
                {
                    return Ok("Nothing found");
                }

                foreach (var income in request.Expense)
                {
                    Expense entity = new()
                    {
                        Category = income.Category,
                        Amount = income.Amount,
                        ExpenseDate = request.Date,
                        Description = income.Description,
                        CreatedBy = _currentUserService.UserId
                    };
                    await _service.Create(entity);
                }

                return Ok();

            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
           
        }


        [HttpPut]
        public async Task<ActionResult> Update([FromBody] AddExpense request)
        {
            if (request.Expense.Count == 0 || String.IsNullOrEmpty(request.Date.ToString()))
            {
                return BadRequest("Invalid Input Data");
            }
            try
            {
                foreach (var income in request.Expense)
                {
                    Expense entity = new()
                    {
                        Id = income.Id,
                        Category = income.Category,
                        Amount = income.Amount,
                        ExpenseDate = request.Date,
                        Description = income.Description,
                        UpdatedDate = DateTime.Now,
                        UpdatedBy = _currentUserService.UserId
                    };
                    await _service.Update(entity);
                }

                return Ok();

            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpDelete("{expenseDate}")]
        public ActionResult Delete(string expenseDate)
        {
            if (String.IsNullOrEmpty(expenseDate.ToString()))
            {
                return BadRequest("Income Date is not selected");
            }
            try
            {
                _service.Delete(expenseDate);
                return Ok();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }

        }

        [HttpGet("{expenseDate}")]
        public ActionResult Get(string expenseDate)
        {
            if(String.IsNullOrEmpty(expenseDate.ToString()))
            {
                return BadRequest("Expense Date is not selected");
            }
            try
            {
                return Ok(_service.Get(expenseDate));
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("list")]
        public ActionResult List([FromQuery] IncomeFilterQueryVm filterQuery)
        {
            var date = DateTime.Now;
            var firstDayOfMonth = new DateTime(date.Year, date.Month, 1);
            var lastDayOfMonth = firstDayOfMonth.AddMonths(1).AddDays(-1);
            filterQuery.StartDate ??= firstDayOfMonth.ToString("yyyy-MM-dd");
            filterQuery.EndDate ??= lastDayOfMonth.ToString("yyyy-MM-dd");

            FilterVm filterVm = new(){
                StartDate = filterQuery.StartDate,
                EndDate = filterQuery.EndDate,
            };
            return Ok(_service.List(filterVm));
        }

        [HttpGet("total-data")]
        public ActionResult TotalData([FromQuery] IncomeFilterQueryVm filterQuery)
        {
            var date = DateTime.Now;
            var firstDayOfMonth = new DateTime(date.Year, date.Month, 1);
            var lastDayOfMonth = firstDayOfMonth.AddMonths(1).AddDays(-1);
            filterQuery.StartDate ??= firstDayOfMonth.ToString("yyyy-MM-dd");
            filterQuery.EndDate ??= lastDayOfMonth.ToString("yyyy-MM-dd");

            FilterVm filterVm = new(){
                StartDate = filterQuery.StartDate,
                EndDate = filterQuery.EndDate,
            };
            return Ok(_service.TotalDataCount(filterVm));
        }
    }
}

File: API/app/Http/Controllers/ExpenseController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use App\Models\Expense;
use Carbon\Carbon;
use Illuminate\Support\Facades\Auth;

class ExpenseController extends Controller
{
     // POST: Create Expenses
    public function create(Request $request)
    {
        try {
            if (empty($request->expense) || empty($request->date)) {
                return response()->json("Nothing found");
            }

            foreach ($request->expense as $expense) {
                Expense::create([
                    'category'     => $expense['category'] ?? null,
                    'amount'       => $expense['amount'] ?? 0,
                    'expense_date' => $request->date,
                    'description'  => $expense['description'] ?? null,
                    'created_by'   => Auth::id(),
                ]);
            }

            return response()->json(['message' => 'Created successfully']);
        } catch (\Exception $ex) {
            return response()->json(['error' => $ex->getMessage()], 400);
        }
    }

    // PUT: Update Expenses
    public function update(Request $request)
    {
        if (empty($request->expense) || empty($request->date)) {
            return response()->json("Invalid Input Data", 400);
        }

        try {
            foreach ($request->expense as $expense) {
                $expenseModel = Expense::find($expense['id']);
                if ($expenseModel) {
                    $expenseModel->update([
                        'category'     => $expense['category'] ?? null,
                        'amount'       => $expense['amount'] ?? 0,
                        'expense_date' => $request->date,
                        'description'  => $expense['description'] ?? null,
                        'updated_by'   => Auth::id(),
                        'updated_at'   => now(),
                    ]);
                }
            }

            return response()->json(['message' => 'Updated successfully']);
        } catch (\Exception $ex) {
            return response()->json(['error' => $ex->getMessage()], 400);
        }
    }

    // DELETE: Delete by date
    public function delete($expenseDate)
    {
        if (empty($expenseDate)) {
            return response()->json("Expense Date is not selected", 400);
        }

        try {
            Expense::where('expense_date', $expenseDate)->delete();
            return response()->json(['message' => 'Deleted successfully']);
        } catch (\Exception $ex) {
            return response()->json(['error' => $ex->getMessage()], 400);
        }
    }

    // GET: Get by date
    public function get($expenseDate)
    {
        if (empty($expenseDate)) {
            return response()->json("Expense Date is not selected", 400);
        }

        try {
            $data = Expense::where('expense_date', $expenseDate)->get();
            return response()->json($data);
        } catch (\Exception $ex) {
            return response()->json(['error' => $ex->getMessage()], 400);
        }
    }

    // GET: List with filter
    public function list(Request $request)
    {
        $date = Carbon::now();
        $firstDayOfMonth = $date->copy()->startOfMonth()->format('Y-m-d');
        $lastDayOfMonth  = $date->copy()->endOfMonth()->format('Y-m-d');

        $startDate = $request->query('start_date', $firstDayOfMonth);
        $endDate   = $request->query('end_date', $lastDayOfMonth);

        $data = Expense::whereBetween('expense_date', [$startDate, $endDate])->get();

        return response()->json($data);
    }

    // GET: Total Expense Data (sum)
    public function total(Request $request)
    {
        $date = Carbon::now();
        $firstDayOfMonth = $date->copy()->startOfMonth()->format('Y-m-d');
        $lastDayOfMonth  = $date->copy()->endOfMonth()->format('Y-m-d');

        $startDate = $request->query('start_date', $firstDayOfMonth);
        $endDate   = $request->query('end_date', $lastDayOfMonth);

        $total = Expense::whereBetween('expense_date', [$startDate, $endDate])->sum('amount');

        return response()->json(['total_expense' => $total]);
    }
}
